lightning/src/ln/channelmanager.rs

```diff
@@ -346,6 +352,14 @@ pub struct ChannelManager<ChanSigner: ChannelKeys> {
 	channel_state: Mutex<ChannelHolder<ChanSigner>>,
 	our_network_key: SecretKey,
 
+	/// Per-peer state storage.
+	/// Because adding or removing an entry is rare, we usually take an outer read lock and then
+	/// operate on the inner value freely. Sadly, this prevents parallel operation when opening a
+	/// new channel.
+	/// If we are connected to a peer we always at least have an entry here, to store their features
+	/// so that we have them available if we open a channel with them and need it for routing.
+	per_peer_state: RwLock<HashMap<PublicKey, Mutex<PerPeerState>>>,
```

Comment:
I was figuring by_id and pending_msg_events (at least for those message events that are non-broadcast, but I dont think there's harm in having broadcast events in there too). That would mean our message queue and channel state per-peer would be under a lock, and it would all be per-peer, which I think is the right design. Indeed, we could lock per-channel too but I don't think its worth it because the caller isn't allowed to have two in-flight message calls for a single peer, so we'd almost never be able to be doing something on both channels at once.

Reply:
Hmmm I'm not sure what do we win on concurrency with putting pending_msg_events there, unless we change get_and_clear_pending_msg_events to be per-peer too but it's just going to push the bottleneck to PeerManager and its Mutex<PeerHolder<Descriptor> ?

> but I don't think its worth it because the caller isn't allowed to have two in-flight message calls for a single peer

Could you elaborate ? Because I always thought than the ChannelMessageHandler API was flexible enough so that you could at least call two handle_* at same time (even if it's likely you contend for locks inside them)

Reply:
re: concurrency - the moving of pending_msg_events into per_peer data is more for logical consistency than anything - we *have* to push events onto that queue while holding the lock we took to edit the Channel data. The only way to keep parallelism is to split the pending_msg_events into per_peer_pending_msg_events and include it in the same Mutex that holds the Channel.

re: elaboration: The ChannelMessageHandler API explicitly states "Messages MAY be called in parallel when they originate from different their_node_ids, however they MUST NOT be called in parallel when the two calls have the same their_node_id". Fundamentally the lightning protocol requires in-order message delivery.